torch/__init__.py

```diff
@@ -500,6 +500,7 @@ def use_deterministic_algorithms(mode, *, warn_only=False):
           ``mode='max'``
         * :func:`torch.Tensor.put_` when ``accumulate=False``
         * :func:`torch.Tensor.put_` when ``accumulate=True`` and called on a CUDA tensor
+        * :func:`torch.Tensor.scatter` when ``src`` is a tensor and ``reduce=None``
```

Comment:
actually, this is a good point; the original bug report just reports that cpu/cuda mismatch, not that the devices internally are nondeterministic.

Reply:
If duplicate indices are given, each with different corresponding values, it's unclear which value should be chosen

Reply:
That makes sense -- but do we know which it is?

Reply:
Turns out it is nondeterministic at least on CUDA:

```python
for n in range(1, 256):
    indices = torch.zeros(n).long()
    last_result = None
    results = []

    for _ in range(10):
        result = torch.scatter(
            torch.ones(n).long(),
            0,
            indices,
            torch.arange(n))[0].item()

        results.append(result)

    if (torch.tensor(results) != results[0]).any():
        print(f'with n = {n}:')
        print(results)
```

Output:

```
with n = 65:
[64, 16, 16, 16, 16, 16, 16, 16, 16, 16]
with n = 73:
[64, 16, 16, 16, 16, 16, 16, 16, 16, 16]
with n = 97:
[96, 80, 80, 80, 80, 80, 80, 80, 80, 80]
...
with n = 223:
[208, 144, 208, 144, 208, 144, 144, 144, 144, 144]
with n = 226:
[208, 208, 224, 208, 224, 224, 224, 208, 224, 224]
with n = 228:
[208, 224, 224, 208, 208, 208, 224, 224, 224, 224]
```

This seems to be because `TensorIterator::for_each` breaks up the operation into multiple parallelized sub-operations if the input size is greater than 64. If two sub-operations write to the same location, the order isn't guaranteed.

Looking at the code, it seems possible that CPU has this same issue. But it is apparently much more regular than the CUDA impl, because I'm having trouble exercising it.

Reply:
I think this is good enough and we should just mark it nondeterministic just to be safe.